core/src/test/java/org/apache/calcite/test/InterpreterTest.java

```diff
@@ -301,6 +301,50 @@ private static void assertRows(Interpreter interpreter,
     final Interpreter interpreter = new Interpreter(dataContext, convert);
     assertRows(interpreter, "[0]", "[10]", "[20]", "[30]");
   }
+
+  @Test public void testInterpretIntersect() throws Exception {
+    final String sql = "select * from\n"
+        + "(select x, y from (values (1, 'a'), (1, 'a'), (2, 'b'), (3, 'c')) as t(x, y))\n"
+        + "intersect\n"
+        + "(select x, y from (values (1, 'a'), (2, 'c'), (4, 'x')) as t2(x, y))\n";
+    SqlNode validate = planner.validate(planner.parse(sql));
+    RelNode convert = planner.rel(validate).rel;
+    final Interpreter interpreter = new Interpreter(dataContext, convert);
+    assertRows(interpreter, "[1, a]");
+  }
+
+  @Test public void testInterpretIntersectAll() throws Exception {
+    final String sql = "select * from\n"
+        + "(select x, y from (values (1, 'a'), (1, 'a'), (2, 'b'), (3, 'c')) as t(x, y))\n"
+        + "intersect all\n"
+        + "(select x, y from (values (1, 'a'), (2, 'c'), (4, 'x')) as t2(x, y))\n";
+    SqlNode validate = planner.validate(planner.parse(sql));
+    RelNode convert = planner.rel(validate).rel;
+    final Interpreter interpreter = new Interpreter(dataContext, convert);
+    assertRows(interpreter, "[1, a]", "[1, a]");
+  }
+
+  @Test public void testInterpretMinus() throws Exception {
+    final String sql = "select * from\n"
+        + "(select x, y from (values (1, 'a'), (2, 'b'), (2, 'b'), (3, 'c')) as t(x, y))\n"
+        + "except\n"
+        + "(select x, y from (values (1, 'a'), (2, 'c'), (4, 'x')) as t2(x, y))\n";
+    SqlNode validate = planner.validate(planner.parse(sql));
+    RelNode convert = planner.rel(validate).rel;
+    final Interpreter interpreter = new Interpreter(dataContext, convert);
+    assertRows(interpreter, "[2, b]", "[3, c]");
+  }
+
+  @Test public void testInterpretMinusAll() throws Exception {
+    final String sql = "select * from\n"
+        + "(select x, y from (values (1, 'a'), (2, 'b'), (2, 'b'), (3, 'c')) as t(x, y))\n"
+        + "except all\n"
+        + "(select x, y from (values (1, 'a'), (2, 'c'), (4, 'x')) as t2(x, y))\n";
+    SqlNode validate = planner.validate(planner.parse(sql));
+    RelNode convert = planner.rel(validate).rel;
+    final Interpreter interpreter = new Interpreter(dataContext, convert);
+    assertRows(interpreter, "[2, b]", "[2, b]", "[3, c]");
```

Comment:
can you add an case for `(2, 'b'), (2, 'b') except (2, 'b')` and  `(2, 'b'), (2, 'b') except all (2, 'b')`?

Reply:
sure, edge case should be covered.
Updated.

Reply:
These testXxx methods have a lot of code in common. Refactor?

Reply:
OK, updated.